src/carousel/tp-carousel.ts

```diff
@@ -0,0 +1,413 @@
+/**
+ * Internal dependencies.
+ */
+import { TPCarouselArrowElement } from './tp-carousel-arrow';
+import { TPCarouselSlideElement } from './tp-carousel-slide';
+import { TPCarouselSlidesElement } from './tp-carousel-slides';
+
+/**
+ * TP Carousel.
+ */
+export class TPCarouselElement extends HTMLElement {
+	protected _observer: IntersectionObserver;
+	protected slides: NodeListOf<TPCarouselSlideElement> | null;
+	protected slideTrack: TPCarouselSlidesElement | null;
+	protected isProgramaticScroll: boolean = false;
+
+	/**
+	 * Constructor.
+	 */
+	constructor() {
+		// Initialize parent.
+		super();
+		this.slides = this.querySelectorAll( 'tp-carousel-slide' );
+		this.slideTrack = this.querySelector( 'tp-carousel-slides' );
+		this._observer = new IntersectionObserver( this.attributeChangeOnScroll?.bind( this ), { root: this.slideTrack, threshold: 0.999 } );
+
+		// Set current slide.
+		if ( ! this.getAttribute( 'current-slide' ) ) {
+			this.setAttribute( 'current-slide', '1' );
+		}
+
+		// Initialize carousel.
+		this.slide();
+		this.setAttribute( 'initialized', 'yes' );
+
+		// Observe which slide is in view.
+		this.slides.forEach( ( slide ) => this._observer.observe( slide ) );
+	}
+
+	/**
+	 * Connected callback.
+	 */
+	connectedCallback() {
+		// Update on initial render.
+		this.update();
+	}
+
+	/**
+	 * Change current-slide attribute on scroll.
+	 *
+	 * @param {IntersectionObserverEntry[]} entries slides which enter or leave on scroll.
+	 */
+	attributeChangeOnScroll( entries: IntersectionObserverEntry[] ): void {
+		// If the scroll is programatic.
+		if ( this.isProgramaticScroll ) {
+			// Do nothing.
+			return;
+		}
+
+		// Change the current slide index when slide comes into view.
+		entries?.forEach( ( entry ) => {
+			// Check if the entry is intersecting with the slide track.
+			if ( entry.isIntersecting && entry.target instanceof TPCarouselSlideElement && this.slides ) {
+				const index = Array.from( this.slides ).indexOf( entry.target );
+
+				// Update current slide index based on if it is is right or left scroll.
+				if ( index + 1 - ( this.perView - 1 ) > this.currentSlideIndex ) {
```

Comment:
@harshdeep-gill Let's hold this into a variable above to make it more readable. 
`index + 1 - ( this.perView - 1 )`